# Choropleth Map of Swiss Grants by Canton

In [1]:
import pandas as pd
import requests
import time
import pickle
import numpy as np
import folium

## Part 1: Creating the Dictionaries mapping Universities/Institutions to Cantons

In [55]:
#Use only University, Institution and Approved Amount
df = pd.read_csv('P3_GrantExport.csv',sep=';', usecols=[6,7,13]) # delimiter is ;
df = df[(df.University.notnull())&(df["Approved Amount"]!='0.00')]

#Nicht zuteilbar is NA in German
df = df[df['University'] != 'Nicht zuteilbar - NA'] 

In [3]:
dict_unis = pickle.load( open( "dict_unis.p", "rb" ) )
dict_private = pickle.load( open( "dict_private.p", "rb" ) )
dict_nop = pickle.load( open( "dict_nop.p", "rb" ) )

In [6]:
gmapsAPIurl = 'https://maps.googleapis.com/maps/api/geocode/json'
gmapsAPIkey = 'AIzaSyClUHUWJ3ZyRkYN3DbilaEGGcAtgvHKsdY'


In [7]:
#list of unique universities
universities = df_unis.University.unique()

In [8]:
# Returns a dict for all universities we can map using google maps API for the list 'universities'
def get_data(universities):
    state = []
    null_names = []
    names = []
    dictionary = {}
    count_good = 0
    count_bad = 0
    for university in universities:
        time.sleep(0.3)
        university_a = university.split(" -",1)[0]
        url = gmapsAPIurl
        parameters = {'address': university_a,'region':'CH','key':gmapsAPIkey}
        r = requests.get(url, params=parameters)
        status=r.json()['status']
        results = r.json()['results']
        if status=='OK':
            flag=1
            count_good=count_good+1
            names.append(university)
            address=[]
            for result in results:
                for address_component in result['address_components']:
                    if flag==1 and address_component['types']==['administrative_area_level_1', 'political']:
                        address.append(address_component['short_name'])
                        print(university_a,":", address )
                        dictionary[university] = address[0]
                        state.append(address[0])
                        flag=0
                        break

        if status=='ZERO_RESULTS':
            count_bad=count_bad+1
            null_names.append(university)
    
    return(dictionary, state, null_names, names, count_good, count_bad)



In [9]:
#run get_data for list of unique univerities
dict_unis, state_unis, null_names_unis, names_unis, count_good_unis, count_bad_unis = get_data(universities)

Université de Genève : ['GE']
Universität Basel : ['BS']
Université de Fribourg : ['FR']
Universität Zürich : ['ZH']
Université de Lausanne : ['VD']
Universität Bern : ['BE']
ETH Zürich : ['ZH']
Universität St. Gallen : ['SG']
Pädagogische Hochschule Graubünden : ['GR']
EPF Lausanne : ['VD']
Pädagogische Hochschule Zürich : ['ZH']
Universität Luzern : ['LU']
Robert Walser-Stiftung Bern : ['BE']
Paul Scherrer Institut : ['AG']
Pädagogische Hochschule St. Gallen : ['SG']
Physikal.-Meteorolog. Observatorium Davos : ['GR']
AO Research Institute : ['Wien']
Kantonsspital St. Gallen : ['SG']
Institut für Kulturforschung Graubünden : ['GR']
Interkant. Hochschule für Heilpädagogik ZH : ['ZH']
Pädagogische Hochschule Bern : ['BE']
Hochschule Luzern : ['LU']
Haute école pédagogique du canton de Vaud : ['VD']
Haute école pédagogique BE, JU, NE : ['NE']
Pädagogische Hochschule Luzern : ['LU']
Istituto Svizzero di Roma : ['Lazio']
Pädag. Hochschule Tessin (Teilschule SUPSI) : ['TI']
Pädagogische Hoc

In [10]:
#Total number of columns
uni_count = df_unis.count().University

In [11]:
uni_count

48314

In [12]:
#Gets the percentage of rows for which we found the canton
def getPercentFound(df, null_names, uniORinsti):
    t = df.copy()
    t['count'] = t.groupby([uniORinsti])[uniORinsti].transform('count')
    t = t[[ uniORinsti, 'count']]
    t = t.drop_duplicates().sort('count')
    t['Found_boolean'] = t[uniORinsti].apply( lambda x: (0 if x in null_names else 1) )
    t['Found'] = t.apply( lambda row: (row['Found_boolean']*row['count']), axis=1)
    sumis = t.sum()
    return(sumis.Found*100/uni_count)

In [13]:
#Run it for the uni mapping we have now
getPercentFound(df_unis, null_names_unis, 'University')

/home/kirtan/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:6: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)


86.109616260297216

In [14]:
# Private institutions and NOP(Non profit organisations) could be in any canton so we need to look at
# institutions for them

df_institutions = df_unis[((df_unis['Institution'].notnull()) & ((df_unis['University']=='NPO (Biblioth., Museen, Verwalt.) - NPO') | (df_unis['University']=='Firmen/Privatwirtschaft - FP')))]
del df_institutions['Approved Amount']

##Private institutions
df_private = df_institutions[df_institutions['University']==('Firmen/Privatwirtschaft - FP')]
df_private1 = df_private['Institution']
private = df_private1.drop_duplicates()

##NOP
df_nop = df_institutions[df_institutions['University']==('NPO (Biblioth., Museen, Verwalt.) - NPO')]
df_nop1 = df_nop['Institution']
nop = df_nop1.drop_duplicates()


In [15]:
# get data for list of institutions in private
dict_private, state_private, null_names_private, names_private, count_good_private, count_bad_private = get_data(private)

Stöckli A. & J. AG : ['AG']
Département de Physique Université de Fribourg : ['FR']
Groupe Limnoceane Institut de géologie Université de Neuchâtel : ['NE']
INFRAS : ['AP']
Siemens Building Technologies AG : ['AG']
Suiselectra Ingenieurunternehmung AG : ['AG']
Prognos AG : ['AG']
Cellulose Attisholz AG : ['AG']
Corporate Communication Von Roll Management AG : ['AG']
HCP Planen und Beraten für das Gesundheitswesen AG : ['AG']
Health Management Institute : ['Jiangsu Sheng']
MOR Informatik AG : ['AG']
Colenco Power Consulting AG : ['AG']
Planconsult W+B AG : ['AG']
Ernst Basler + Partner AG : ['AG']
Aqua-System AG : ['AG']
Interdisziplinäre Berater- und Forschungsgruppe AG : ['AG']
HOLDERBANK Management und Beratung AG : ['AG']
Aarproject AG : ['AG']
Gruppe CORSO : ['LU']
Gesundheitsdepartement des Kantons Basel-Stadt : ['BS']
Ateliers et Laboratoire C R E P H A R T : ['Grand Casablanca']
Forschungszentrum ABB Schweiz AG : ['AG']
Ciba-Geigy AG Forschungszentrum Marly : ['FR']
FEINTOOL AG :

In [16]:
# get data for list of institutions in nop
dict_nop, state_nop, null_names_nop, names_nop, count_good_nop, count_bad_nop = get_data(nop)

Schweizerisches Nationalmuseum Landesmuseum Zürich : ['ZH']
Historische und Antiquarische Gesellschaft zu Basel : ['BS']
Forum Ost-West : ['HE']
Büro für Archäologie der Stadt Zürich : ['ZH']
Kantonsschule Sargans : ['SG']
Staatsarchiv : ['NDS']
Kuratorium der Helvetia Sacra c/o Staatsarchiv Basel : ['BS']
Archäologische Bodenforschung des Kantons Basel-Stadt : ['BS']
Département des travaux publics du Canton de Vaud : ['VD']
Ecole Suisse d'archéologie en Grèce  Univ. de Lausanne : ['VD']
Kommission des Phonogrammarchivs Universität Zürich : ['ZH']
Staatsarchiv Uri : ['UR']
Tschudi-Kommission der AGGS Staatsarchiv Zürich : ['ZH']
Archives d'Etat de Genève : ['GE']
Bibliothek Kantonsschule Luzern : ['LU']
Musées d'art et d'histoire de Genève : ['GE']
Kantonsschule Baden : ['AG']
Museum des Kantons Thurgau : ['TG']
Pro Helvetia : ['BE']
Liceo di Lugano 2 : ['TI']
Laténium Parc et musée d'archéologie de Neuchâtel : ['NE']
Archives de l'Etat de Fribourg : ['FR']
Gymnase du Bugnon : ['Auver

In [17]:
count_good_nop

146

In [18]:
#Doesn't work perfectly for institutions
getPercentFound(df_nop, null_names_nop, 'Institution')

/home/kirtan/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:6: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)


0.93968621931531238

In [22]:
# Save a dictionary into a pickle file.
# Don't need to run this again!

# pickle.dump( dict_unis, open( "dict_unis.p", "wb" ) )
# pickle.dump( dict_private, open( "dict_private.p", "wb" ) )
# pickle.dump( dict_nop, open( "dict_nop.p", "wb" ) )

## Part 2: Creating the Map

Steps into creating the map include:
* Mapping Dictionary into Canton Column in DataFrame
    * university
    * private
    * nop

**Cleaning the DataFrame**
* Manually add missing ones to get 95% overall
    * Checking the completeness of Cantons
    * Manual Mappings of University to Canton
* Removing Non-Swiss Cantons
* Aggregating DataFrame into Canton and Amount only
* Adding Missing Cantons

** Using Folium to map DataFrame into Choropleth **
* Map the DataFrame into the Chloropeth
    * Fine tune legend

### Mapping Dictionary into Canton Column in DataFrame

In [15]:
# mapping University and Insitutions to cantons
df["Canton"] = df["University"].map(dict_unis) # universities
df.ix[df.University ==('Firmen/Privatwirtschaft - FP'), 'Canton'] = df[df["University"] ==('Firmen/Privatwirtschaft - FP')]['Institution'].map(dict_private) # private
df.ix[df.University ==('NPO (Biblioth., Museen, Verwalt.) - NPO'), 'Canton'] = df[df["University"] ==('NPO (Biblioth., Museen, Verwalt.) - NPO')]['Institution'].map(dict_nop) # nop

### Filling 95% of Universities with Canton information

**Checking the completeness of Cantons**

Using the Google Maps API, we were able to map 87.4% of the Universities/Institutions.

In [8]:
print("Percentage Filled:",(df.Canton.shape[0]-df[pd.isnull(df['Canton'])].shape[0])/df.Canton.shape[0])

Percentage Filled 0.874115163306702


In [19]:
df.Canton.shape[0]

48314

In [21]:
df[pd.isnull(df['Canton'])].shape[0]

6082

** Manual Mappings of University to Canton **

Total # of Universities is `df.Canton.shape[0]` which is 48314.
$$ 0.05*48314 = 2415 $$

Currently 6082 are missing values, so we need to fill out at least $$6089-2415 ~= 3674$$ Universities. To choose the most frequent Universities, we use the dataframe below as a reference.

We added the rest manually to fill out over 95% of Universities. There are a total 

To get to 95% mapping, we need to add 2415 missing, means need $$6089-2415 ~= 3674$$ more maps. These insitutions, like WSL or HSE(?) have many locations, so we chose the headquarters.

In [24]:
# This table is used to show the Universities with most entries in amount, thus most important to map
df[pd.isnull(df['Canton'])].groupby('University').count().sort_values(by='Approved Amount', ascending=False).head(20)

Institution  \
University                                                        
Université de Neuchâtel - NE                               1559   
NPO (Biblioth., Museen, Verwalt.) - NPO                     935   
Università della Svizzera italiana - USI                    345   
Eidg. Anstalt für Wasserversorgung - EAWAG                  331   
Firmen/Privatwirtschaft - FP                                296   
HES de Suisse occidentale - HES-SO                          271   
Zürcher Fachhochschule (ohne PH) - ZFH                      259   
Eidg. Material und Prüfungsanstalt - EMPA                   235   
Fachhochschule Nordwestschweiz (ohne PH) - FHNW             226   
Eidg. Forschungsanstalt für Wald,Schnee,Land - WSL          221   
Inst. de Hautes Etudes Internat. et du Dév - IHEID          189   
Berner Fachhochschule - BFH                                 136   
Forschungsanstalten Agroscope - AGS                         135   
SUP della Svizzera italiana - SUPSI                         135   
Idiap Research Institute - IDIAP                             87   
Friedrich Miescher Institute - FMI                           84   
Weitere Spitäler - ASPIT                                     81   
Allergie- und Asthmaforschung - SIAF                         44   
Weitere Institute - FINST                                    43   
Inst. universit. romand de Santé au Travail - IST            35   

                                                    Approved Amount  Canton  
University                                                                   
Université de Neuchâtel - NE                                   1593       0  
NPO (Biblioth., Museen, Verwalt.) - NPO                        1020       0  
Università della Svizzera italiana - USI                        346       0  
Eidg. Anstalt für Wasserversorgung - EAWAG                      333       0  
Firmen/Privatwirtschaft - FP                                    312       0  
HES de Suisse occidentale - HES-SO                              271       0  
Zürcher Fachhochschule (ohne PH) - ZFH                          260       0  
Eidg. Material und Prüfungsanstalt - EMPA                       238       0  
Fachhochschule Nordwestschweiz (ohne PH) - FHNW                 228       0  
Eidg. Forschungsanstalt für Wald,Schnee,Land - WSL              221       0  
Inst. de Hautes Etudes Internat. et du Dév - IHEID              198       0  
Berner Fachhochschule - BFH                                     136       0  
Forschungsanstalten Agroscope - AGS                             135       0  
SUP della Svizzera italiana - SUPSI                             135       0  
Idiap Research Institute - IDIAP                                 87       0  
Friedrich Miescher Institute - FMI                               84       0  
Weitere Spitäler - ASPIT                                         81       0  
Allergie- und Asthmaforschung - SIAF                             44       0  
Weitere Institute - FINST                                        43       0  
Inst. universit. romand de Santé au Travail - IST                35       0

In [25]:
# extra universities manually mapped
dict_extra = {'Université de Neuchâtel - NE': 'NE',
               'Università della Svizzera italiana - USI': 'TI',
               'Eidg. Anstalt für Wasserversorgung - EAWAG': 'ZH',
               'HES de Suisse occidentale - HES-SO': 'JU',
               'Zürcher Fachhochschule (ohnet PH) - ZFH': 'ZH',
               'Eidg. Material und Prüfungsanstalt - EMPA': 'ZH',
               'Fachhochschule Nordwestschweiz (ohne PH) - FHNW': 'GS',
               'Eidg. Forschungsanstalt für Wald,Schnee,Land - WSL': 'ZH',
               'Inst. de Hautes Etudes Internat. et du Dév - IHEID': 'GE',
               'Berner Fachhochschule - BFH': 'BE',
               'Forschungsanstalten Agroscope - AGS': 'ZH',
               'SUP della Svizzera italiana - SUPSI': 'TI',
               }

In [97]:
# mapping the Universities with Cantons missing
df.ix[pd.isnull(df.Canton), 'Canton'] = df[pd.isnull(df.Canton)]['University'].map(dict_extra)

In [98]:
# the percentage filled after manual map
print("Percentage Filled",(df.Canton.shape[0]-df[pd.isnull(df['Canton'])].shape[0])/df.Canton.shape[0])

Percentage Filled 0.9534710435898497


### Removing Non-Swiss Cantons

We need to remove non-Swiss Cantons.

In [99]:
df.Canton.unique()

array(['GE', nan, 'BS', 'FR', 'ZH', 'VD', 'BE', 'NE', 'HE', 'SG', 'NDS',
       'GR', 'UR', 'LU', 'AG', 'TG', 'TI', 'JU', 'Auvergne-Rhône-Alpes',
       'Bourgogne Franche-Comté', 'ZG', 'Pichincha', 'SH',
       'Languedoc-Roussillon-Midi-Pyrénées', 'VS', 'NRW', 'GP', 'Wien',
       'AP', 'Jiangsu Sheng', 'Grand Casablanca', 'WB', 'GS', 'SO',
       'Mie Prefecture', 'England', 'Nordwest', 'Hiroshima-ken',
       'Kareliya Republits', 'Guangdong', 'BC', 'Sichuan Sheng', 'DL',
       'Van', 'OW', 'Shida Kartli', 'Saitama-ken', 'BL', 'Lazio', 'Edo',
       'SZ', 'Chocó'], dtype=object)

In [49]:
#Keep only the swiss cantons, remove cantons that are incorrect
if not not df_unis_canton.empty:
    df_unis_canton=df.copy()
df=df_unis_canton[(df_unis_canton.Canton.str[0]>=chr(65)) & (df_unis_canton.Canton.str[0]<=chr(90)) & 
                              (df_unis_canton.Canton.str[1]>=chr(65)) & (df_unis_canton.Canton.str[1]<=chr(90)) & 
                              (df_unis_canton.Canton!='NDS') & (df_unis_canton.Canton!='NRW') & (df_unis_canton.Canton!='AP') &
                              (df_unis_canton.Canton!='BC') & (df_unis_canton.Canton!='CA') & (df_unis_canton.Canton!='DL') & 
                              (df_unis_canton.Canton!='HE') & (df_unis_canton.Canton!='WB') & (df_unis_canton.Canton!='GP')]

In [50]:
df.sort_values(by='Canton').Canton.unique()

array(['AG', 'BE', 'BL', 'BS', 'FR', 'GE', 'GR', 'LU', 'NE', 'OW', 'SG',
       'SH', 'SO', 'SZ', 'TG', 'TI', 'UR', 'VD', 'VS', 'ZG', 'ZH'], dtype=object)

### Aggregating DataFrame into Canton and Amount only

In [31]:
# cleaning dataframe and building the grouped by DataFrame

if 'Approved Amount' in df.columns: # we change the name 'Approved Amount to Amount', avoid running code more than once
    df['Approved Amount'] = pd.to_numeric(df['Approved Amount'], errors='coerce')
    df = pd.DataFrame(df.groupby('Canton').sum()['Approved Amount'].apply(np.round))
    df.reset_index(inplace=1) # what I've been looking for
    df.columns = ['Canton', 'Amount']

### Adding Missing Cantons

In [37]:
missing_cantons = {'Canton' : ['NW', 'GL', 'AR', 'AI', 'JU'],
                   'Amount' : np.zeros(5)}

In [38]:
missing_cantons = pd.DataFrame(missing_cantons)

In [52]:
df = pd.concat([df,missing_cantons])

In [53]:
# DataFrame with Canton and Total Funding
df.sort_values(by='Amount')
df['Amount'] = np.log10(df['Amount']) # log
df.fillna(0)

Amount Canton
0  -0.040113     AG
4  -0.016264     BE
5  -0.134771     BL
6  -0.017385     BS
12 -0.027949     FR
13 -0.014605     GE
15 -0.067421     GR
22 -0.051405     LU
27 -0.091442     NE
30 -0.148144     OW
32 -0.045596     SG
33 -0.109996     SH
34 -0.111309     SO
35 -0.110090     SZ
39 -0.079322     TG
40 -0.086394     TI
41 -0.154402     UR
42 -0.012350     VD
43 -0.114830     VS
47 -0.107569     ZG
48 -0.008930     ZH
0       -inf     NW
1       -inf     GL
2       -inf     AR
3       -inf     AI
4       -inf     JU

### Building the Map with Folium

In [108]:
cantons_geo = r'ch-cantons.topojson.json'

swiss_map = folium.Map(location=[47, 8],tiles='cartodbpositron', zoom_start=7)
swiss_map.geo_json(geo_path=cantons_geo, 
                     data=df_unis_canton_total,
                     columns=['Canton', 'Amount'],
                     key_on='feature.id',
                     topojson='objects.cantons',
                     fill_color='YlGn', fill_opacity=0.75, line_opacity=0.2,
                     legend_name = 'Funding Amount',
                     threshold_scale=[0, 5, 6, 7, 8, 9] #Scale depends on the maximum and minimum value in "Amount"
                    )
swiss_map.save('swiss_cantons.html')


/Users/tbfang/anaconda3/lib/python3.5/site-packages/folium/folium.py:504: UserWarning: This method is deprecated. Please use Map.choropleth instead.
  warnings.warn('This method is deprecated. '


In [109]:
swiss_map

## Part 3: Bonus
BONUS: using the map you have just built, and the geographical information contained in it, could you give a rough estimate of the difference in research funding between the areas divided by the Röstigraben?

* Map Cantons into German-speaking and French-speaking
* Do a statistical test

### French vs German Speaking Region

Bern is a special case. G/F

In [54]:
df.sort_values(by='Canton').Canton.unique()

array(['AG', 'AI', 'AR', 'BE', 'BL', 'BS', 'FR', 'GE', 'GL', 'GR', 'JU',
       'LU', 'NE', 'NW', 'OW', 'SG', 'SH', 'SO', 'SZ', 'TG', 'TI', 'UR',
       'VD', 'VS', 'ZG', 'ZH'], dtype=object)

In [ ]:
# extra universities manually mapped
dict_rostigraben = {'AG': 'G', 'AI': 'G', 'AR': 'G',
                    'BE': 'G', 'BL': 'G', 'BS': 'G',
                    'GE': 'F', 'GL': 'G', 'GR': 'G',
                    'JU': 'F', 'LU': 'G', 'NE': 'G',
                    'NW': 'F', 'OW': 'G', 'SG': 'G',
                    'SH': 'F', 'SO': 'G', 'SZ': 'G',
                    'TG': 'F', 'TI': 'G', 'UR': 'G',
                    'VD': 'F', 'VS': 'G', 'ZG': 'G',
                    'ZH': 'F', 'FR': 'G'}